### PR TITLE
Simplified Stellarium integration and added subscriber sign-up to the Streamlit Dashboard. 

### LINKED ISSUE #52: Create a subscribe form python file for the dashboard
As an amateur astronomer, I want to have a way to sign up to receive alerts based on my location, so that if my area is likely to be able to see something in the night sky, I will receive either a text or email.

### PR DESCRIPTION
**Pull request by MurasatoNaoya**
Simplified the integration of the Stellarium iFrame by only setting the default location to London and removing webpage customisations that I can't get to work - I will ask Dan about this later in the week. 

A subscriber sign-up form has also been implemented, allowing users to register with either an email or UK phone number, with validation for both formats in the form of RegEx.

This closes #52

### CODE REVIEW

File: dashboard/dashboard.py [hunk 1]
```diff
@@ -1,106 +1,157 @@
...
+import psycopg2
 from datetime import datetime
-import urllib.parse
-from streamlit_autorefresh import st_autorefresh
-
-
-st.title("⭐ Starwatch Data Dashboard ⭐")
-
-
-st.markdown("""
```
Comment (AlexOBarnes) [inline dashboard/dashboard.py:10]: Would be good to add our logo now

File: dashboard/dashboard.py [hunk 2]
```diff
@@ -1,106 +1,157 @@
...
+        (?:(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+ 
+        [a-z0-9](?:[a-z0-9-]*[a-z0-9])?           
+        |                                
+        \[(?:(?:(2(5[0-5]|[0-4][0-9])|1[0-9][0-9]
+        |[1-9]?[0-9]))\.){3}
+        (?:(2(5[0-5]|[0-4][0-9])|1[0-9][0-9]
+        |[1-9]?[0-9])|[a-z0-9-]*[a-z0-9]:       
+        (?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21-\x5a\x53-\x7f]
+        |\\[\x01-\x09\x0b\x0c\x0e-\x7f])+)\])  
+    '''
```
Comment (AlexOBarnes) [inline dashboard/dashboard.py:45]: Good to add checks for subscribers

File: dashboard/dashboard.py [hunk 3]
```diff
@@ -1,106 +1,157 @@
...
+                with connect_to_db() as conn:
+                    with conn.cursor() as cursor:
+                        cursor.execute('''
+                            INSERT INTO subscriber (subscriber_username, subscriber_phone, subscriber_email) 
+                            VALUES (%s, %s, %s);
+                        ''', (username, phone if phone else None, email if email else None)) # Using tertiary statements to account for None(s).
+                    conn.commit()  # Commiting the valid user information to the 'subscriber' table in the RDS.
+                st.success('Subscriber added successfully!')
+            except Exception as e:
+                st.error(f'Error adding subscriber: {e}')
```
Comment (AlexOBarnes) [inline dashboard/dashboard.py:157]: Good use of error handling

File: dashboard/dashboard.py [hunk 4]
```diff
@@ -1,106 +1,157 @@
...
+        }
+        query_string = urllib.parse.urlencode(params)
+        full_url = f'{base_url}?{query_string}'
+        return full_url
+
+    def get_current_datetime():
+        return datetime.now()
+
+    # The Stellarium iFrame is stored in a Streamlit container.
+    with st.container():
```
Comment (AlexOBarnes) [inline dashboard/dashboard.py:103]: Useful for separating out the code

File: dashboard/dashboard.py [hunk 5]
```diff
@@ -1,106 +1,157 @@
...
+    DEFAULT_LONGITUDE = -0.1278
+
+    # Unless a specific location is provided, London used as a default.
+    if 'latitude' not in st.session_state:
+        st.session_state['latitude'] = DEFAULT_LATITUDE
+    if 'longitude' not in st.session_state:
+        st.session_state['longitude'] = DEFAULT_LONGITUDE
+
+
+    def generate_stellarium_url(lat, lon, dt):
```
Comment (AlexOBarnes) [inline dashboard/dashboard.py:87]: Very good integration

**Review by AlexOBarnes — APPROVED** [review]
Thanks for simplifying this

File: dashboard/dashboard.py (hunk 1 above)
Comment (CalebMathers) [reply dashboard/dashboard.py:10]: A logo would be very snazzy

File: dashboard/dashboard.py [hunk 6]
```diff
@@ -1,106 +1,157 @@
...
-
-with st.container():
-    st.markdown("## Interactive Night Sky View")
+def validate_email(email):
+    '''
+    Validates an email address using an RFC 5322 compliant regex.
+    Returns a re match object if valid, None if invalid.
+    '''
+
+    email_regex = r'''
```
Comment (CalebMathers) [inline dashboard/dashboard.py:29]: Very in-depth regex, nice

File: dashboard/dashboard.py [hunk 7]
```diff
@@ -1,106 +1,157 @@
...
-*This dashboard automatically updates every minute to provide the most current view of the night sky.*
-""")
\ No newline at end of file
+    - With the +44 country code: +44 followed by 10 digits (e.g., +44 1234567890)
+    - Without the country code: 0 followed by 10 digits (e.g., 01234567890
+
+    Returns a re match object if valid, None if invalid.
+    '''
+
+    phone_regex = r'^(\+44\d{10}|0\d{10})$'
```
Comment (CalebMathers) [inline dashboard/dashboard.py:60]: Good phone regex for UK numbers, which is all we need for now and could always be changed in the future when StarWatch takes over the world

File: dashboard/dashboard.py [hunk 8]
```diff
@@ -1,106 +1,157 @@
...
+        current_datetime = get_current_datetime()
+        
+        stellarium_url = generate_stellarium_url(
+            DEFAULT_LATITUDE, 
+            DEFAULT_LONGITUDE, 
+            current_datetime
+        )
+        
+        st.components.v1.html(f'''
+        <iframe src='{stellarium_url}' style='width: 100%; height: 600px; border: none;'></iframe>
```
Comment (CalebMathers) [inline dashboard/dashboard.py:116]: Yoink

File: dashboard/dashboard.py [hunk 9]
```diff
@@ -1,106 +1,157 @@
...
+    username = st.text_input('Desired Username', '')
+    email = st.text_input('Email Address', '')
+    phone = st.text_input('Phone Number (UK)', '')
+
+    if st.button('Submit'):
+        # Ensure username, either email or phone is provided
+        if not username:
+            st.error('Please provide a username.')
+
+        elif not email and not phone:
```
Comment (CalebMathers) [inline dashboard/dashboard.py:136]: Good checks to make sure it won't violate any constraints in the database

File: dashboard/dashboard.py [hunk 10]
```diff
@@ -1,106 +1,157 @@
...
+        else:
+            try:
+                with connect_to_db() as conn:
+                    with conn.cursor() as cursor:
+                        cursor.execute('''
+                            INSERT INTO subscriber (subscriber_username, subscriber_phone, subscriber_email) 
+                            VALUES (%s, %s, %s);
+                        ''', (username, phone if phone else None, email if email else None)) # Using tertiary statements to account for None(s).
+                    conn.commit()  # Commiting the valid user information to the 'subscriber' table in the RDS.
+                st.success('Subscriber added successfully!')
```
Comment (CalebMathers) [inline dashboard/dashboard.py:155]: Feedback to the user, good

**Review by CalebMathers — APPROVED** [review]

File: dashboard/dashboard.py (hunk 7 above)
Comment (MurasatoNaoya) [reply dashboard/dashboard.py:60]: StarWatch going worldwide.

**Pull request merged**